File: rvm-main/src/whichkey.lua
-- src/whichkey.lua
-- LazyVim-style Which-Key popup.
-- When the user presses <leader> (default: Space), we display a grouped menu.
-- Two-key chords: <leader>ff (find files), <leader>fg (live grep), etc.
--
-- This module is the registry + render logic. It is consumed by `keybindings.lua`.

local WhichKey = {}
WhichKey.__index = WhichKey

-- ─────────────────────────────────────────────────────────
-- Registry: top-level groups (single char shown when leader pressed)
-- Each group has: label, icon, entries = { [chord] = { label = ..., action = ... } }
-- ─────────────────────────────────────────────────────────
WhichKey.groups = {
  f = { label = "Find",   icon = "󰍉", desc = "Telescope-like finders" },
  g = { label = "Git",    icon = "󰊢", desc = "Git workflow" },
  b = { label = "Buffers", icon = "󰓩", desc = "Buffer operations" },
  p = { label = "Projects", icon = "󰉋", desc = "Project management" },
  l = { label = "LSP",    icon = "󰌵", desc = "Language server" },
  x = { label = "Diagnostics", icon = "󰅙", desc = "Diagnostics panel" },
  t = { label = "Terminal", icon = "󰆍", desc = "Terminal panel" },
  e = { label = "Explorer", icon = "󰉋", desc = "Toggle file explorer" },
  w = { label = "Window",  icon = "󰖲", desc = "Window management" },
  s = { label = "Search",  icon = "󰍉", desc = "Buffer search" },
  q = { label = "Quit",    icon = "󰗼", desc = "Quit / session" },
  h = { label = "Help",    icon = "󰋽", desc = "Help" },
  d = { label = "Debug",   icon = "󰃤", desc = "Debug" },
}

-- ─────────────────────────────────────────────────────────
-- Chord registry: { "ff" = {label, action}, "fg" = ..., ... }
-- ─────────────────────────────────────────────────────────
WhichKey.chords = {
  -- File finders (Telescope-like)
  ["ff"] = { label = "Find Files",        action = "finder_files",     group = "f" },
  ["fg"] = { label = "Live Grep",         action = "finder_grep",       group = "f" },
  ["fb"] = { label = "Buffers",           action = "finder_buffers",    group = "f" },
  ["fr"] = { label = "Recent Files",      action = "finder_recent",     group = "f" },
  ["fc"] = { label = "Commands",          action = "finder_commands",   group = "f" },
  ["fh"] = { label = "Help",              action = "finder_help",       group = "f" },
  ["fk"] = { label = "Key Maps",          action = "finder_keymaps",    group = "f" },

  -- Git
  ["gg"] = { label = "Git Status",        action = "git_status",        group = "g" },
  ["gd"] = { label = "Git Diff",          action = "git_diff",          group = "g" },
  ["gc"] = { label = "Git Commit",        action = "git_commit",        group = "g" },
  ["gp"] = { label = "Git Push",          action = "git_push",          group = "g" },
  ["gl"] = { label = "Git Log",           action = "git_log",            group = "g" },
  ["gb"] = { label = "Git Branch",        action = "git_branch",        group = "g" },
  ["ga"] = { label = "Git Stage All",     action = "git_stage_all",     group = "g" },
  ["gu"] = { label = "Git Unstage All",   action = "git_unstage_all",   group = "g" },

  -- Buffers
  ["bd"] = { label = "Delete Buffer",     action = "buffer_delete",    group = "b" },
  ["bo"] = { label = "Close Other Buffers", action = "buffer_close_others", group = "b" },
  ["bn"] = { label = "Next Buffer",       action = "buffer_next",       group = "b" },
  ["bp"] = { label = "Previous Buffer",   action = "buffer_prev",       group = "b" },
  ["bl"] = { label = "Buffer List",       action = "buffer_list",        group = "b" },

  -- Projects
  ["pp"] = { label = "Project Switcher",  action = "project_switch",    group = "p" },
  ["pf"] = { label = "Project Files",     action = "project_files",     group = "p" },
  ["pg"] = { label = "Project Grep",      action = "project_grep",      group = "p" },
  ["pr"] = { label = "Recent Projects",   action = "project_recent",    group = "p" },
  ["pd"] = { label = "Project Dashboard", action = "project_dashboard", group = "p" },

  -- LSP
  ["la"] = { label = "Code Actions",      action = "lsp_code_action",   group = "l" },
  ["lr"] = { label = "Rename Symbol",     action = "lsp_rename",        group = "l" },
  ["lf"] = { label = "Format Buffer",     action = "lsp_format",        group = "l" },
  ["lh"] = { label = "Hover",              action = "lsp_hover",          group = "l" },
  ["ld"] = { label = "Definition",        action = "lsp_definition",    group = "l" },
  ["ll"] = { label = "References",        action = "lsp_references",    group = "l" },
  ["ls"] = { label = "Signature Help",    action = "lsp_signature",     group = "l" },
  ["lw"] = { label = "Workspace Symbols", action = "lsp_workspace_syms", group = "l" },

  -- Diagnostics
  ["xx"] = { label = "Diagnostics Panel", action = "diagnostics_panel",  group = "x" },
  ["xn"] = { label = "Next Diagnostic",  action = "diagnostics_next",   group = "x" },
  ["xp"] = { label = "Prev Diagnostic",  action = "diagnostics_prev",   group = "x" },
  ["xo"] = { label = "Open Diagnostics", action = "diagnostics_open",   group = "x" },

  -- Terminal
  ["tt"] = { label = "Toggle Terminal",  action = "terminal_toggle",   group = "t" },
  ["tf"] = { label = "Terminal Float",   action = "terminal_float",    group = "t" },
  ["to"] = { label = "Open New Terminal", action = "terminal_new",      group = "t" },

  -- Window
  ["ws"] = { label = "Split Horizontal",  action = "window_split",      group = "w" },
  ["wv"] = { label = "Split Vertical",    action = "window_vsplit",      group = "w" },
  ["wc"] = { label = "Close Window",      action = "window_close",      group = "w" },
  ["wo"] = { label = "Close Others",      action = "window_close_others", group = "w" },
  ["wh"] = { label = "Focus Left",        action = "window_focus_left", group = "w" },
  ["wj"] = { label = "Focus Down",        action = "window_focus_down", group = "w" },
  ["wk"] = { label = "Focus Up",          action = "window_focus_up",   group = "w" },
  ["wl"] = { label = "Focus Right",       action = "window_focus_right", group = "w" },
  ["w="] = { label = "Equalize",          action = "window_equalize",   group = "w" },
  ["w-"] = { label = "Decrease Height",   action = "window_dec_h",      group = "w" },
  ["w+"] = { label = "Increase Height",    action = "window_inc_h",      group = "w" },
  ["w<"] = { label = "Decrease Width",    action = "window_dec_w",      group = "w" },
  ["w>"] = { label = "Increase Width",    action = "window_inc_w",      group = "w" },

  -- Single-key shortcuts (no chord, just one key after leader)
  ["e"]  = { label = "Toggle Explorer",   action = "explorer_toggle",   single = true },
  ["E"]  = { label = "Reveal Current File", action = "explorer_reveal",  single = true },
  ["w"]  = { label = "Save File",         action = "file_save",         single = true },
  ["q"]  = { label = "Quit",              action = "app_quit",          single = true },
  ["Q"]  = { label = "Force Quit",        action = "app_force_quit",    single = true },
  ["n"]  = { label = "New Buffer",        action = "buffer_new",        single = true },
  ["t"]  = { label = "Cycle Themes",      action = "theme_cycle",      single = true },
  ["s"]  = { label = "Cycle Styles",      action = "style_cycle",      single = true },
  ["u"]  = { label = "Undo",              action = "edit_undo",         single = true },
  ["r"]  = { label = "Redo",              action = "edit_redo",         single = true },
  ["i"]  = { label = "Icons Mode",        action = "icons_toggle",      single = true },
}

-- ─────────────────────────────────────────────────────────
-- State machine
-- ─────────────────────────────────────────────────────────
function WhichKey.new()
  local self = setmetatable({}, WhichKey)
  self.visible        = false
  self.prefix         = ""          -- e.g. "f" once user pressed <leader>f
  self.last_result    = nil        -- filled when user completes a chord
  return self
end

function WhichKey:open()
  self.visible = true
  self.prefix  = ""
end

function WhichKey:close()
  self.visible = false
  self.prefix  = ""
  self.last_result = nil
end

function WhichKey:is_open()
  return self.visible
end

-- Feed a key. Returns: ("chord", chord_action) | ("prefix", prefix_char) | ("cancel", nil) | ("noop", nil)
function WhichKey:feed(key)
  if not self.visible then return "noop", nil end
  if key == "esc" or key == " " then
    if self.prefix == "" and key == " " then
      -- User pressed Space twice → cancel
      self:close()
      return "cancel", nil
    end
    self:close()
    return "cancel", nil
  end

  local new_prefix = self.prefix .. key

  -- Check for single-key chord
  if self.prefix == "" then
    local entry = WhichKey.chords[key]
    if entry and entry.single then
      self:close()
      return "chord", entry.action
    end
  end

  -- Check for two-key chord
  if #new_prefix == 2 then
    local entry = WhichKey.chords[new_prefix]
    if entry then
      self:close()
      return "chord", entry.action
    end
    -- Not found, cancel
    self:close()
    return "cancel", nil
  end

  -- One key so far — check if it's a known group prefix
  if #new_prefix == 1 then
    if WhichKey.groups[new_prefix] then
      self.prefix = new_prefix
      return "prefix", new_prefix
    end
    -- Unknown single key, but check single-key chords again
    local entry = WhichKey.chords[new_prefix]
    if entry and entry.single then
      self:close()
      return "chord", entry.action
    end
    self:close()
    return "cancel", nil
  end

  self:close()
  return "cancel", nil
end

-- ─────────────────────────────────────────────────────────
-- Render hint data — returns a list of { key=, label=, icon= } rows
-- UI layer actually draws them.
-- ─────────────────────────────────────────────────────────
function WhichKey:hint_rows()
  if not self.visible then return {} end

  if self.prefix == "" then
    -- Show top-level groups + single-key chords
    local rows = {}
    for k, g in pairs(WhichKey.groups) do
      table.insert(rows, { key = k, label = g.label, icon = g.icon, desc = g.desc })
    end
    -- Single-key shortcuts
    for k, e in pairs(WhichKey.chords) do
      if e.single then
        table.insert(rows, { key = k, label = e.label, icon = "*", desc = "" })
      end
    end
    -- Sort: groups first (alphabetical), then singles (alphabetical)
    table.sort(rows, function(a, b) return a.key < b.key end)
    return rows
  else
    -- Show entries in current group prefix
    local rows = {}
    for chord, e in pairs(WhichKey.chords) do
      if #chord == 2 and chord:sub(1, 1) == self.prefix then
        table.insert(rows, { key = chord:sub(2, 2), label = e.label, icon = "›" })
      end
    end
    table.sort(rows, function(a, b) return a.key < b.key end)
    return rows
  end
end

return WhichKey

File: src/whichkey.rs
// src/whichkey.rs — LazyVim-style Which-Key popup for the Rust binary.
//
// Per spec requirement #16: When pressing Space, show:
//   f  Find
//   g  Git
//   b  Buffers
//   p  Projects
//   l  LSP
//   x  Diagnostics
//   t  Terminal
//   e  Explorer
//
// This module is the *registry* + *state machine*; rendering lives in `ui.rs`.

#[derive(Debug, Clone)]
pub struct WhichKeyEntry {
    pub key: char,
    pub label: String,
    pub icon: String,
    pub desc: String,
}

#[derive(Debug, Clone, Default)]
pub struct WhichKey {
    pub visible: bool,
    pub prefix: String,
    pub last_action: Option<String>,
}

impl WhichKey {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn open(&mut self) {
        self.visible = true;
        self.prefix.clear();
    }

    pub fn close(&mut self) {
        self.visible = false;
        self.prefix.clear();
        self.last_action = None;
    }

    pub fn is_open(&self) -> bool {
        self.visible
    }

    /// Feed a key. Returns: ("chord", action) | ("prefix", prefix) | ("cancel", nil) | ("noop", nil)
    pub fn feed(&mut self, key: char) -> (&'static str, Option<String>) {
        if !self.visible {
            return ("noop", None);
        }
        if key == '\x1b' || key == ' ' {
            self.close();
            return ("cancel", None);
        }

        // Single-key shortcuts
        let single_action = match key {
            'e' => Some("explorer_toggle"),
            'w' => Some("file_save"),
            'q' => Some("app_quit"),
            'n' => Some("buffer_new"),
            't' => Some("theme_cycle"),
            's' => Some("style_cycle"),
            'u' => Some("edit_undo"),
            'r' => Some("edit_redo"),
            _ => None,
        };
        if self.prefix.is_empty() && single_action.is_some() {
            let action = single_action.unwrap().to_string();
            self.close();
            return ("chord", Some(action));
        }

        let new_prefix = format!("{}{}", self.prefix, key);

        // Check for two-key chord
        if new_prefix.len() == 2 {
            let action = Self::chord_action(&new_prefix);
            self.close();
            return if let Some(a) = action {
                ("chord", Some(a.to_string()))
            } else {
                ("cancel", None)
            };
        }

        // One-key prefix
        if new_prefix.len() == 1 {
            if Self::is_group_prefix(&new_prefix) {
                self.prefix = new_prefix;
                return ("prefix", Some(self.prefix.clone()));
            }
            self.close();
            return ("cancel", None);
        }

        self.close();
        ("cancel", None)
    }

    /// Get the hint rows for rendering.
    pub fn hint_rows(&self) -> Vec<WhichKeyEntry> {
        let mut rows = Vec::new();

        if self.prefix.is_empty() {
            // Top-level groups
            let groups = [
                ('f', "Find",       "󰍉", "Telescope-like finders"),
                ('g', "Git",         "󰊢", "Git workflow"),
                ('b', "Buffers",     "󰓩", "Buffer operations"),
                ('p', "Projects",    "󰉋", "Project management"),
                ('l', "LSP",         "󰌵", "Language server"),
                ('x', "Diagnostics", "󰅙", "Diagnostics panel"),
                ('t', "Terminal",    "󰆍", "Terminal panel"),
                ('e', "Explorer",    "󰉋", "Toggle file explorer"),
                ('w', "Window",      "󰖲", "Window management"),
                ('s', "Search",      "󰍉", "Buffer search"),
            ];
            for (k, label, icon, desc) in groups {
                rows.push(WhichKeyEntry {
                    key: k,
                    label: label.to_string(),
                    icon: icon.to_string(),
                    desc: desc.to_string(),
                });
            }
        } else {
            // Show entries in current group prefix
            let chords = Self::chords_for_prefix(&self.prefix);
            for (k, label) in chords {
                rows.push(WhichKeyEntry {
                    key: k,
                    label: label.to_string(),
                    icon: "›".to_string(),
                    desc: String::new(),
                });
            }
        }
        rows
    }

    fn is_group_prefix(p: &str) -> bool {
        matches!(p, "f" | "g" | "b" | "p" | "l" | "x" | "t" | "w" | "s")
    }

    fn chord_action(chord: &str) -> Option<&'static str> {
        match chord {
            // File finders
            "ff" => Some("finder_files"),
            "fg" => Some("finder_grep"),
            "fb" => Some("finder_buffers"),
            "fr" => Some("finder_recent"),
            "fc" => Some("finder_commands"),
            // Git
            "gg" => Some("git_status"),
            "gd" => Some("git_diff"),
            "gc" => Some("git_commit"),
            "gp" => Some("git_push"),
            "gl" => Some("git_log"),
            // Buffers
            "bd" => Some("buffer_delete"),
            "bo" => Some("buffer_close_others"),
            "bn" => Some("buffer_next"),
            "bp" => Some("buffer_prev"),
            // Projects
            "pp" => Some("project_switch"),
            "pf" => Some("project_files"),
            "pg" => Some("project_grep"),
            "pd" => Some("project_dashboard"),
            // LSP
            "la" => Some("lsp_code_action"),
            "lr" => Some("lsp_rename"),
            "lf" => Some("lsp_format"),
            "lh" => Some("lsp_hover"),
            "ld" => Some("lsp_definition"),
            "ll" => Some("lsp_references"),
            // Diagnostics
            "xx" => Some("diagnostics_panel"),
            "xn" => Some("diagnostics_next"),
            "xp" => Some("diagnostics_prev"),
            // Terminal
            "tt" => Some("terminal_toggle"),
            // Window
            "ws" => Some("window_split"),
            "wv" => Some("window_vsplit"),
            "wc" => Some("window_close"),
            "wh" => Some("window_focus_left"),
            "wj" => Some("window_focus_down"),
            "wk" => Some("window_focus_up"),
            "wl" => Some("window_focus_right"),
            _ => None,
        }
    }

    fn chords_for_prefix(prefix: &str) -> Vec<(char, &'static str)> {
        match prefix {
            "f" => vec![
                ('f', "Find Files"),
                ('g', "Live Grep"),
                ('b', "Buffers"),
                ('r', "Recent Files"),
                ('c', "Commands"),
            ],
            "g" => vec![
                ('g', "Git Status"),
                ('d', "Git Diff"),
                ('c', "Git Commit"),
                ('p', "Git Push"),
                ('l', "Git Log"),
            ],
            "b" => vec![
                ('d', "Delete Buffer"),
                ('o', "Close Other Buffers"),
                ('n', "Next Buffer"),
                ('p', "Previous Buffer"),
            ],
            "p" => vec![
                ('p', "Project Switcher"),
                ('f', "Project Files"),
                ('g', "Project Grep"),
                ('d', "Project Dashboard"),
            ],
            "l" => vec![
                ('a', "Code Actions"),
                ('r', "Rename Symbol"),
                ('f', "Format Buffer"),
                ('h', "Hover"),
                ('d', "Definition"),
                ('l', "References"),
            ],
            "x" => vec![
                ('x', "Diagnostics Panel"),
                ('n', "Next Diagnostic"),
                ('p', "Prev Diagnostic"),
            ],
            "t" => vec![
                ('t', "Toggle Terminal"),
                ('f', "Terminal Float"),
                ('o', "Open New Terminal"),
            ],
            "w" => vec![
                ('s', "Split Horizontal"),
                ('v', "Split Vertical"),
                ('c', "Close Window"),
                ('h', "Focus Left"),
                ('j', "Focus Down"),
                ('k', "Focus Up"),
                ('l', "Focus Right"),
            ],
            _ => vec![],
        }
    }
}
